File: admin/internal/model/schema.go
// Copyright © 2026-present reepu.com
// SPDX-License-Identifier: Apache-2.0
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ContentSchemaKind 内容模型类型
type ContentSchemaKind string

const (
	ContentSchemaKindCollection ContentSchemaKind = "collection" // 集合类型（多条目）
	ContentSchemaKindSingle    ContentSchemaKind = "single"      // 单条类型
)

// ContentSchema 内容模型
type ContentSchema struct {
	ID                   uuid.UUID        `json:"id" gorm:"type:uuid;primaryKey;default:gen_random_uuid()"`
	SiteID               uuid.UUID        `json:"site_id" gorm:"type:uuid;not null;index"`
	Name                 string           `json:"name" gorm:"size:200;not null"`
	Slug                 string           `json:"slug" gorm:"size:100;not null;index"`
	Description          string           `json:"description" gorm:"type:text"`
	Kind                 ContentSchemaKind  `json:"kind" gorm:"type:schema_kind;not null;default:'collection'"`
	VersioningEnabled    bool             `json:"versioning_enabled" gorm:"default:false"`
	DraftAutosaveInterval *int            `json:"draft_autosave_interval" gorm:"default:null"`
	IsActive             bool             `json:"is_active" gorm:"default:true"`
	SortOrder            int              `json:"sort_order" gorm:"default:0"`
	DataSignature       string           `json:"-" gorm:"type:varchar(256);not null;default:''"` // 防篡改签名（HMAC-SHA256 hex）
	CreatedBy            *uuid.UUID       `json:"created_by" gorm:"type:uuid"`
	CreatedTime            time.Time        `json:"created_time" gorm:"autoCreateTime"`
	UpdatedTime            time.Time        `json:"updated_time" gorm:"autoUpdateTime"`
	DeletedAt              gorm.DeletedAt    `json:"deleted_time" gorm:"column:deleted_time;index"`

	// 关联
	Fields []Field `json:"fields,omitempty" gorm:"foreignKey:schema_id;references:ID"`
}

// TableName 表名
func (ContentSchema) TableName() string {
	return "contful_schemas"
}

// Field 字段定义
type Field struct {
	ID                  uuid.UUID  `json:"id" gorm:"type:uuid;primaryKey;default:gen_random_uuid()"`
	ContentSchemaID       uuid.UUID  `json:"schema_id" gorm:"column:schema_id;type:uuid;not null;index"`
	Name                string     `json:"name" gorm:"size:100;not null"`
	Label               string     `json:"label" gorm:"size:200;not null"`
	Description         string     `json:"description" gorm:"type:text"`
	FieldType           string     `json:"field_type" gorm:"type:field_type;not null"`
	Config              JSONB      `json:"config" gorm:"type:jsonb;default:'{}'"`
	Validation          JSONB      `json:"validation" gorm:"type:jsonb;default:'{}'"`
	Display             JSONB      `json:"display" gorm:"type:jsonb;default:'{}'"`
	DefaultValue        *JSONB     `json:"default_value" gorm:"type:jsonb"`
	SortOrder           int        `json:"sort_order" gorm:"default:0"`
	ConditionalDisplay  *JSONB      `json:"conditional_display" gorm:"type:jsonb"`
	CreatedTime           time.Time  `json:"created_time" gorm:"autoCreateTime"`
	UpdatedTime           time.Time  `json:"updated_time" gorm:"autoUpdateTime"`
	DeletedAt              gorm.DeletedAt `json:"deleted_time" gorm:"column:deleted_time;index"`
}

// TableName 表名
func (Field) TableName() string {
	return "contful_fields"
}

// FieldType 字段类型枚举
var FieldTypes = []string{
	"text",      // 单行文本
	"rich_text", // 富文本
	"number",    // 数字
	"boolean",   // 布尔值
	"date",      // 日期
	"datetime",  // 日期时间
	"email",     // 邮箱
	"url",       // URL（网页链接 / 媒体文件地址 / 图片 URL）
	"json",      // JSON
	"relation",  // 关联（指向其他内容模型）
	"enum",      // 枚举
	"password",  // 密码
}

File: console/src/api/schema.ts
// Copyright © 2026-present reepu.com
// SPDX-License-Identifier: Apache-2.0

import { get, post, put, del } from '@/utils/request'

// ============ ContentSchema API ============

export interface ContentSchema {
  id: string
  site_id: string
  name: string
  slug: string
  description: string
  kind: 'collection' | 'single'
  versioning_enabled: boolean
  draft_autosave_interval: number | null
  is_active: boolean
  sort_order: number
  created_by: string | null
  created_time: string
  updated_time: string
  fields?: Field[]
}

export interface Field {
  id: string
  schema_id: string
  name: string
  label: string
  description: string
  field_type: FieldType
  config: Record<string, any>
  validation: Record<string, any>
  display: Record<string, any>
  default_value: any
  sort_order: number
  conditional_display: any
  created_time: string
  updated_time: string
}

export type FieldType =
  | 'text'
  | 'rich_text'
  | 'number'
  | 'boolean'
  | 'date'
  | 'datetime'
  | 'email'
  | 'url'
  | 'json'
  | 'media'
  | 'relation'
  | 'enum'
  | 'password'

export const FIELD_TYPES: Record<FieldType, { labelKey: string; icon: string }> = {
  text: { labelKey: 'fieldTypes.text', icon: 'text' },
  rich_text: { labelKey: 'fieldTypes.richText', icon: 'richtext' },
  number: { labelKey: 'fieldTypes.number', icon: 'number' },
  boolean: { labelKey: 'fieldTypes.boolean', icon: 'boolean' },
  date: { labelKey: 'fieldTypes.date', icon: 'calendar' },
  datetime: { labelKey: 'fieldTypes.dateTime', icon: 'calendar' },
  email: { labelKey: 'fieldTypes.email', icon: 'mail' },
  url: { labelKey: 'fieldTypes.url', icon: 'link' },
  json: { labelKey: 'fieldTypes.json', icon: 'code' },
  media: { labelKey: 'fieldTypes.media', icon: 'image' },
  relation: { labelKey: 'fieldTypes.relation', icon: 'relation' },
  enum: { labelKey: 'fieldTypes.enum', icon: 'list' },
  password: { labelKey: 'fieldTypes.password', icon: 'lock' },
}

export interface ContentSchemaCreate {
  name: string
  slug: string
  description?: string
  kind: 'collection' | 'single'
  versioning_enabled?: boolean
  draft_autosave_interval?: number
  sort_order?: number
}

export interface ContentSchemaUpdate {
  name?: string
  slug?: string
  description?: string
  kind?: 'collection' | 'single'
  versioning_enabled?: boolean
  draft_autosave_interval?: number
  is_active?: boolean
  sort_order?: number
}

export interface FieldCreate {
  name: string
  label: string
  description?: string
  field_type: FieldType
  config?: Record<string, any>
  validation?: Record<string, any>
  display?: Record<string, any>
  default_value?: any
  sort_order?: number
  conditional_display?: Record<string, any>
}

export interface FieldUpdate {
  name?: string
  label?: string
  description?: string
  field_type?: FieldType
  config?: Record<string, any>
  validation?: Record<string, any>
  display?: Record<string, any>
  default_value?: any
  sort_order?: number
  conditional_display?: Record<string, any>
}

export interface ListResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

// 创建内容模型
export function createContentSchema(data: ContentSchemaCreate) {
  return post<ContentSchema>('/content/schemas', data)
}

// 获取内容模型列表
export function getContentSchemas(params?: { page?: number; page_size?: number }) {
  return get<ListResponse<ContentSchema>>('/content/schemas', { params })
}

// 获取内容模型详情
export function getContentSchema(id: string) {
  return get<ContentSchema>(`/content/schemas/${id}`)
}

// 更新内容模型
export function updateContentSchema(id: string, data: ContentSchemaUpdate) {
  return put<ContentSchema>(`/content/schemas/${id}`, data)
}

// 删除内容模型
export function deleteContentSchema(id: string) {
  return del(`/content/schemas/${id}`)
}

// ============ Field API ============

// 创建字段
export function createField(schemaId: string, data: FieldCreate) {
  return post<Field>(`/content/schemas/${schemaId}/fields`, data)
}

// 获取字段列表
export function getFields(schemaId: string) {
  return get<{ items: Field[] }>(`/content/schemas/${schemaId}/fields`)
}

// 更新字段
export function updateField(schemaId: string, fieldId: string, data: FieldUpdate) {
  return put<Field>(`/content/schemas/${schemaId}/fields/${fieldId}`, data)
}

// 删除字段
export function deleteField(schemaId: string, fieldId: string) {
  return del(`/content/schemas/${schemaId}/fields/${fieldId}`)
}

// 重新排序字段
export function reorderFields(schemaId: string, orders: Record<string, number>) {
  return post(`/content/schemas/${schemaId}/fields/reorder`, { orders })
}
